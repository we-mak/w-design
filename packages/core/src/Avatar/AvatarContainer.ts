import styled from "styled-components";
import { space } from "styled-system";
import { AvatarProps } from "./index";

const getSize = (props: AvatarProps) => {
  switch (props.size) {
    case "xs":
      return `
      height: 1.4rem;
      width: 1.4rem;
      `;
    case "sm":
      return `
      height: 2rem;
      width: 2rem;
      `;
    case "md":
      return `
      height: 2.8rem;
      width: 2.8rem;
      `;
    case "lg":
      return `
      height: 3.2rem;
      width: 3.2rem;
      `;
    case "xl":
      return `
      height: 4rem;
      width: 4rem;
      `;
    case "xxl":
      return `
      height: 6rem;
      width: 6rem;
      `;
    default:
      return `
    height: 1.8rem;
    width: 1.8rem;
    `;
  }
};

export const AvatarContainer = styled.figure<AvatarProps>`
  border-radius: 50%;
  color: rgba(255, 255, 255, 0.85);
  display: inline-block;
  line-height: 1.25;
  margin: 0;
  position: relative;
  vertical-align: middle;
  ${getSize};
  ${space};
`;
