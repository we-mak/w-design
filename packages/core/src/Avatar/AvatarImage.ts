import styled from "styled-components";
import Image from "../Image";

export const AvatarImage = styled(Image)`
  border-radius: 50%;
  height: 100%;
  width: 100%;
  position: relative;
  z-index: 1;
`;
