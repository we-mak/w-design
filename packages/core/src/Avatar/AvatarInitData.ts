import styled from "styled-components";
import { AvatarProps } from "./index";

const getTextSize = (props: AvatarProps) => {
  switch (props.size) {
    case "xs":
      return `font-size: .6rem;`;
    case "sm":
      return `font-size: .9rem;`;
    case "md":
      return `font-size: 1rem;`;
    case "lg":
      return `font-size: 1.8rem;`;
    case "xl":
      return `font-size: 2rem;`;
    case "xxl":
      return `font-size: 2.8rem;`;
    default:
      return `font-size: 1.4rem;`;
  }
};

export const AvatarInitData = styled.span<AvatarProps>`
  color: currentColor;
  left: 50%;
  position: absolute;
  top: 50%;
  transform: translate(-50%, -50%);
  z-index: 1;
  text-transform: uppercase;
  font-weight: 500;
  ${getTextSize};
`;
