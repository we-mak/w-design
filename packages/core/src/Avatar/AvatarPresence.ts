import styled from "styled-components";
import { AvatarProps } from "./index";

const getPresenceStyle = (props: AvatarProps) => {
  switch (props.size) {
    case "xs":
      return `
    height: .4rem;
    width: .4rem;
    right: 0;
    border: 0.075rem solid #fff;
    `;
    case "sm":
      return `
    height: .5rem;
    width: .5rem;
    right: 0;
    border: 0.075rem solid #fff;
    `;
    case "md":
      return `
    height: .7rem;
    width: .7rem;
    right: 0;
    border: 0.075rem solid #fff;
    `;
    case "lg":
      return `
    height: .7rem;
    width: .7rem;
    right: .1rem;
    border: 0.075rem solid #fff;
    `;

    case "xl":
      return `
    height: .8rem;
    width: .8rem;
    right: .3rem;
    border: 0.1rem solid #fff;
    `;
    case "xxl":
      return `
    height: .9rem;
    width: .9rem;
    right: .9rem;
    border: 0.1rem solid #fff;
    `;
    default:
      return `
    height: .7rem;
    width: .7rem;
    right: .7rem;
    border: 0.1rem solid #fff;
  `;
  }
};

const setPresenceColor = (props: AvatarProps) => {
  const {
    presence,
    theme: { colors }
  } = props;

  if (presence === "online") return colors["G30"];
  if (presence === "offline") return colors["N20"];
  if (presence === "busy") return colors["R30"];

  return colors["G30"];
};

export const AvatarPresence = styled.span<AvatarProps>`
  border-radius: 50%;
  bottom: 0.1rem;
  pointer-events: none;
  position: absolute;
  z-index: 30;
  background-color: ${setPresenceColor};
  ${getPresenceStyle};
`;
