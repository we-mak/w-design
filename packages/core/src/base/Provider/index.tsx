import * as React from "react";
import { ThemeProvider } from "styled-components";
import defaultTheme from "../../common/theme";
import { GlobalStyle } from "./GlobalStyle";
import "./normalize.css";

interface ThemeProps extends React.HTMLProps<HTMLDivElement> {
  theme?: object | ((theme: object) => object);
}

function Provider(p: ThemeProps) {
  const { theme = {}, children } = p;

  return (
    <ThemeProvider theme={{ ...defaultTheme, ...theme }}>
      <>
        <GlobalStyle />
        {children}
      </>
    </ThemeProvider>
  );
}

Provider.displayName = "WProvider";

export default Provider;
