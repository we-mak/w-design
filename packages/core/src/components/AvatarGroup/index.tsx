import * as React from "react";
import styled from "styled-components";
import { AvatarGroupProps } from "./types";

const AvatarGroupWrapper = styled.div`
  align-content: space-between;
  align-items: flex-start;
  display: flex;
`;
AvatarGroupWrapper.displayName = "AvatarGroupWrapper";

const AvatarWrapper = styled.div`
  flex: 0 0 auto;
`;
AvatarWrapper.displayName = "AvatarWrapper";

const AvatarContent = styled.div`
  flex: 1 1 auto;
  align-self: center;

  &:not(:last-child) {
    padding-right: 0.4rem;
  }
  &:not(:first-child) {
    padding-left: 0.4rem;
  }
`;
AvatarContent.displayName = "AvatarContent";

const AvatarTitle = styled.div`
  font-weight: 500;
  margin-bottom: 0.2em;
`;
AvatarTitle.displayName = "AvatarTitle";

const AvatarDescription = styled.div`
  color: ${props => props.theme.colors["N40"]};
  font-size: ${props => props.theme.fontSizes[0]};
`;

const AvatarActions = styled.div`
  flex: 0 0 auto;
`;

const AvatarGroup = ({
  avatar,
  name,
  description,
  actions,
  ...rest
}: AvatarGroupProps) => {
  return (
    <AvatarGroupWrapper {...rest}>
      <AvatarWrapper>{avatar}</AvatarWrapper>
      <AvatarContent>
        {name && <AvatarTitle>{name}</AvatarTitle>}
        {description && <AvatarDescription>{description}</AvatarDescription>}
      </AvatarContent>
      {actions && <AvatarActions>{actions}</AvatarActions>}
    </AvatarGroupWrapper>
  );
};

export default AvatarGroup;
export { AvatarGroupProps };
