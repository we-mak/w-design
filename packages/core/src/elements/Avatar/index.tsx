import * as React from "react";
import {
  AvatarWrapper,
  AvatarDataInit,
  AvatarImage,
  PresenceMark
} from "./Styled";
import { firstLetter, randomProperty, filterObject } from "@w-design/helpers";
import { AvatarProps } from "./types";
import { colors } from "../../common/colors";

const Avatar = ({
  size = "md",
  avatarUrl,
  dataInitial = "We mak",
  alt,
  presence,
  ...rest
}: AvatarProps) => {
  const [backgroundColor, setBackgroundColor] = React.useState("");

  React.useEffect(() => {
    if (!avatarUrl) {
      setBackgroundColor(
        randomProperty(filterObject(colors, ["WHITE", "BLACK", "N1"]))
      );
    }
  }, []);

  return (
    <AvatarWrapper style={{ backgroundColor }} size={size} {...rest}>
      {dataInitial && (
        <AvatarDataInit size={size}>{firstLetter(dataInitial)}</AvatarDataInit>
      )}
      {avatarUrl && <AvatarImage data-src={avatarUrl} alt={alt} />}
      {presence && <PresenceMark presence={presence} size={size} />}
    </AvatarWrapper>
  );
};

export default React.memo(Avatar);
export { AvatarProps };
