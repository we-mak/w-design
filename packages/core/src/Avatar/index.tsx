import * as React from "react";
import { firstLetter, randomProperty, filterObject } from "@w-design/helpers";
import { colors } from "../common/colors";
import { GlobProps } from "../common/props";
import { AvatarContainer } from "./AvatarContainer";
import { AvatarPresence } from "./AvatarPresence";
import { AvatarInitData } from "./AvatarInitData";
import { AvatarImage } from "./AvatarImage";

export interface AvatarProps extends GlobProps {
  size?: "xs" | "sm" | "md" | "lg" | "xl" | "xxl";
  // User online status
  presence?: "online" | "offline" | "busy";
  avatarUrl?: string;
  // Initial content of avatar if without image
  dataInitial?: string;
  alt?: string;
  crossOrigin?: React.ImgHTMLAttributes<any>["crossOrigin"];
}

const Avatar = ({
  size = "md",
  avatarUrl,
  dataInitial = "We mak",
  alt,
  crossOrigin,
  presence,
}: AvatarProps) => {
  const [backgroundColor, setBackgroundColor] = React.useState("white");

  React.useEffect(() => {
    if (!avatarUrl) {
      setBackgroundColor(
        randomProperty(filterObject(colors, ["WHITE", "BLACK", "N1"]))
      );
    }
  }, []);

  return (
    <AvatarContainer style={{ backgroundColor }} size={size}>
      {dataInitial && (
        <AvatarInitData size={size}>{firstLetter(dataInitial)}</AvatarInitData>
      )}
      {avatarUrl && (
        <AvatarImage data-src={avatarUrl} alt={alt} crossOrigin={crossOrigin} />
      )}
      {presence && <AvatarPresence presence={presence} size={size} />}
    </AvatarContainer>
  );
};

export default React.memo(Avatar);
