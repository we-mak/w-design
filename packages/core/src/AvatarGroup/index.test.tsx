import * as React from "react";
import { render, cleanup } from "@testing-library/react";
import Provider from "../Provider";
import Avatar from "../Avatar";
import AvatarGroup from "./index";

afterEach(cleanup);

describe("<AvatarGroup/>", () => {
  // Render test
  it("should renders correctly", () => {
    const container = render(
      <Provider>
        <AvatarGroup
          avatar={<Avatar avatarUrl="https://api.adorable.io/avatars/80/vn" />}
          name="we-design"
          description="this is the description"
          actions={<div>do action</div>}
        />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });
});
