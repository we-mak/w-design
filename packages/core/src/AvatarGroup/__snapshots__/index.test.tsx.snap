// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AvatarGroup/> should renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c0 {
  -webkit-align-content: space-between;
  -ms-flex-line-pack: space-between;
  align-content: space-between;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
}

.c5 {
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

.c5:not(:last-child) {
  padding-right: 0.4rem;
}

.c5:not(:first-child) {
  padding-left: 0.4rem;
}

.c6 {
  font-weight: 500;
  margin-bottom: 0.2em;
}

.c7 {
  color: #4D6171;
  font-size: 0.7rem;
}

.c8 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
}

.c2 {
  border-radius: 50%;
  color: rgba(255,255,255,0.85);
  display: inline-block;
  line-height: 1.25;
  margin: 0;
  position: relative;
  vertical-align: middle;
  height: 2.8rem;
  width: 2.8rem;
}

.c3 {
  color: currentColor;
  left: 50%;
  position: absolute;
  top: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  z-index: 1;
  text-transform: uppercase;
  font-weight: 500;
  font-size: 1rem;
}

.c4 {
  border-radius: 50%;
  height: 100%;
  width: 100%;
  position: relative;
  z-index: 1;
}

<body>
    <div>
      <div
        class="c0"
      >
        <div
          class="c1"
        >
          <figure
            class="c2"
          >
            <span
              class="c3"
            >
              Wm
            </span>
            <img
              class="c4"
              data-src="https://api.adorable.io/avatars/80/vn"
              src="data:image/gif;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mP88R8AAvUB+VkkrXoAAAAASUVORK5CYII="
            />
          </figure>
        </div>
        <div
          class="c5"
        >
          <div
            class="c6"
          >
            we-design
          </div>
          <div
            class="c7"
          >
            this is the description
          </div>
        </div>
        <div
          class="c8"
        >
          <div>
            do action
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-AykKC BhJoj"
    >
      <div
        class="sc-AykKD csiNKD"
      >
        <figure
          class="sc-AykKK dgcyCa"
        >
          <span
            class="sc-fzXfLO iUOMKx"
          >
            Wm
          </span>
          <img
            class="sc-AykKI bFthup sc-fzXfLP kfgWgc"
            data-src="https://api.adorable.io/avatars/80/vn"
            src="data:image/gif;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mP88R8AAvUB+VkkrXoAAAAASUVORK5CYII="
          />
        </figure>
      </div>
      <div
        class="sc-AykKE ikrQef"
      >
        <div
          class="sc-AykKF ktpJRT"
        >
          we-design
        </div>
        <div
          class="sc-AykKG gLpGPU"
        >
          this is the description
        </div>
      </div>
      <div
        class="sc-AykKH fsIIiH"
      >
        <div>
          do action
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
